### PR TITLE
TOOLS-2013: Support PKCS #8 encrypted client private keys

### PR DESCRIPTION
**Pull request by edobranov**
Jira: https://jira.mongodb.org/browse/TOOLS-2013

This adds the ability to decrypt PKCS 8 encrypted PEM keys.

The assumption is that these types of keys are the only format possible in PEM blocks labeled `-----BEGIN ENCRYPTED PRIVATE KEY-----`. I wasn't able to find any place that explicitly states this, though the [PKCS 8 documentation](https://tools.ietf.org/html/rfc5208) mentions an `EncryptedPrivateKeyInfo` type, which looks to correspond to that label. So I believe it's unique to PKCS 8.

There's one caveat that I'll make a comment about below.

### CODE REVIEW

File: db/db.go [hunk 1]
```diff
@@ -189,14 +190,35 @@ func addClientCertFromBytes(cfg *tls.Config, data []byte, keyPasswd string) (str
...
+					keyBytes, err = x509.DecryptPEMBlock(currentBlock, []byte(keyPasswd))
+					if err != nil {
+						return "", err
+					}
+				} else if strings.Contains(currentBlock.Type, "ENCRYPTED") {
+					// The pkcs8 package only handles the PKCS #5 v2.0 scheme.
+					decrypted, err := pkcs8.ParsePKCS8PrivateKey(currentBlock.Bytes, []byte(keyPasswd))
+					if err != nil {
+						return "", err
+					}
```
Comment (edobranov) [inline db/db.go:213]: From my understanding, there's a couple ways to encrypt a PEM key in PKCS 8 format. The only one that the `pkcs8` Go package can currently handle is called the PKCS 5 v2.0 scheme. The first time I tried this, I accidentally used the wrong scheme and got this error:

`error configuring the connector: error configuring client, can't load client certificate: pkcs8: only PKCS #5 v2.0 supported`

Even though the error is probably specific enough, I'm wondering if we should make a note of this in the docs under the [`--sslPEMKeyFile` entry](https://docs.mongodb.com/database-tools/mongodump/#cmdoption-mongodump-sslpemkeyfile) and provide guidance on getting the right scheme (e.g. using openssl with the `-v2 des3` option to get PKCS 5 v2.0).

I'm okay either way, but curious to hear others' thoughts.

Comment (huan-Mongo) [reply db/db.go:213]: Nice catch! I agree we should add this limitation to the doc so that it won't cause confusion to the client.

**Review by huan-Mongo — APPROVED** [review]
Good job! LGTM!

File: db/db.go [hunk 2]
```diff
@@ -189,14 +190,35 @@ func addClientCertFromBytes(cfg *tls.Config, data []byte, keyPasswd string) (str
 			certDecodedBlock = currentBlock.Bytes
 			start += len(certBlock)
 		} else if strings.HasSuffix(currentBlock.Type, "PRIVATE KEY") {
-			if keyPasswd != "" && x509.IsEncryptedPEMBlock(currentBlock) {
-				var encoded bytes.Buffer
-				buf, err := x509.DecryptPEMBlock(currentBlock, []byte(keyPasswd))
-				if err != nil {
-					return "", err
+			isEncrypted := x509.IsEncryptedPEMBlock(currentBlock) || strings.Contains(currentBlock.Type, "ENCRYPTED")
```
Comment (tfogo) [inline db/db.go:193]: Nit: I think it's better to check for equality against the full "ENCRYPTED PRIVATE KEY" string here. It's better to be precise.

**Review by tfogo — APPROVED** [review]
LGTM.

`addClientCertFromBytes()` is currently replicated in mtc because we are waiting on the go driver to implement GODRIVER-1588. But the eventual goal is to move back to using the version of `addClientCertFromBytes()` that exists in the go driver. So we probably want to open a go driver ticket to add PKCS#8 support there too.

File: Gopkg.toml [hunk 3]
```diff
@@ -48,3 +48,7 @@
   go-tests = true
   unused-packages = true
 
+
+[[constraint]]
+  name = "github.com/youmark/pkcs8"
```
[the remark this answers is not in the document]
Comment (rychipman) [reply Gopkg.toml:53]: @edobranov could you file a ticket for this?

File: db/db.go (hunk 1 above)
Comment (rychipman) [reply db/db.go:213]: I'm not sure that this is a limitation significant enough to warrant a mention in the docs. Generally, the docs describe what we _do_ support, and I think a limitation has to be pretty notable/important for us to want to describe what we _don't_ do.

File: Gopkg.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (edobranov) [reply Gopkg.toml:53]: Yep, filed [TOOLS-2787](https://jira.mongodb.org/browse/TOOLS-2787).

**Comment by edobranov** [comment]
> So we probably want to open a go driver ticket to add PKCS#8 support there too.

There's a Go driver ticket linked to the tools one from a while back ([GODRIVER-354](https://jira.mongodb.org/browse/GODRIVER-364)).

**Comment by tfogo** [comment]
It might be worth talking to the go driver team to see if they think the solution here would work for them and if they're open to you contributing this.

**Comment by rychipman** [comment]
I'm removing myself as a reviewer on this PR, since there are already two other LGTMs, and I don't think my review needs to be a blocker here!

**Pull request merged**